### DISCUSSION TITLE
Formatting lost after base64 conversion

### DISCUSSION THREAD

**Discussion by ajRED777** [dolanmiu/docx#1656] [Q&A] [answered]
The screenshot below shows 2 documents-My Document(37), and My Document(38).
38 was generated using the Express demo app hosted by dolanmiu- https://runkit.com/dolanmiu/docx-demo4 
37 was generated using a sam app with lambda and API gateway locally.
Doc 37 has lost the table formatting but 38 has retained it.
![image](https://user-images.githubusercontent.com/44117428/189820694-a13887db-f06c-4d4d-a979-6cb1ac790292.png)

Code used to generate 37 is as follows:

    const { Document, Packer, Paragraph, Table, TableCell, TableRow } = docx;
    const correlationId = context.awsRequestId;
    
    const table = new Table({
        rows: [
            new TableRow({
                children: [
                    new TableCell({
                        children: [new Paragraph("Hello")],
                    }),
                    new TableCell({
                        children: [],
                    }),
                ],
            }),
            new TableRow({
                children: [
                    new TableCell({
                        children: [],
                    }),
                    new TableCell({
                        children: [new Paragraph("World")],
                    }),
                ],
            }),
        ],
    });
    
    const doc = new Document({
        sections: [{
            children: [table],
        }],
    });
    
    const b64string = await Packer.toBase64String(doc);
    
    return {
        headers: { 'Content-Type': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                    'Content-Disposition': 'attachment; filename=My Document.docx'},
        statusCode: 200,
        body: b64string,
        isBase64Encoded: true
    }

The template building is basically the same as the demo app-https://runkit.com/dolanmiu/docx-demo4 
The only difference is the base64 generation, which is as follows(present in the above code but just putting to show specific change from demo code)
![image](https://user-images.githubusercontent.com/44117428/189823682-d478c59a-9e30-4976-a8f9-97efb1243513.png)

Also tried decoding the base64 string using https://base64.guru/converter/decode/file and got the doc with a broken format similar to 37. (Did this to verify if the AWS API gateway is changing something)

Tried Buffer object also but yielded the same results.
![image](https://user-images.githubusercontent.com/44117428/189829497-f676136c-3323-4ebe-8c10-3ad7e28ea2b5.png)

Hoping someone could help in identifying why the doc does not retain the formatting.

**Comment by ajRED777** [comment]
Apparently, it is not the base 64 conversion that is the issue. It is actually the version of the library that was different- I am using 7.4.1 but the  [demo ](https://runkit.com/dolanmiu/docx-demo4)is using 6.0.1 where something has changed with the margins of a doc.

Just putting here for reference:
If you are developing this using a SAM app with API gateway then you should specify in your response the 'Content-Type' and also the flag 'isBase64Encoded' with 'BinaryMediaType' property for API gateway in yml file as follows:
![image](https://user-images.githubusercontent.com/44117428/190168089-029f68e2-1e16-4f18-83d0-f0ea695518e2.png)
This will ensure the base64 is converted by the API gateway to a binary format and also not garble it into a different binary.


@dolanmiu are there any ways to get the same result in the newer versions i.e. getting margins in the document like in 6.0.1?

**Comment by dolanmiu** [comment, marked as answer]
Could you try explicitly adding margins to the page like so:

https://github.com/dolanmiu/docx/blob/master/demo/6-page-borders.ts
